proposals/XXXX-spirv-resource-representation.md

```diff
@@ -0,0 +1,283 @@
+`RasterizerOrderedStructuredBuffer`, `AppendStructureBuffer`, and
+`ConsumeStructuredBuffer`. These resource types do not map directly to SPIR-V.
+They have multiple implicit features that need to map to different SPIR-V:
+
+1.  They all contain an array of memory that maps to a storage buffer.
+2.  Other than `StructuredBuffer`, they all contain a separate counter variable
+    that is its own storage buffer.
+3.  The references to `RasterizerOrderedStructuredBuffer` are contained in
+    implicit critical regions. In SPIR-V, explicit instructions are used to
+    start and stop the critical region.
+
+This makes it impossible to create a handle type that maps directly to a SPIR-V
+type. To handle this, we will create a target type `spirv.VulkanBuffer`:
+
+```
+target("spirv.VulkanBuffer", ElementType, StorageClass, IsWriteable, IsROV)
+```
+
+`ElementType` is the type for the storage buffer array, and `StorageClass` is
+the storage class for the array. `IsWriteable` is true if the resource can be
+written to, and `IsROV` is true if it is a rasterizer order view. If the
+resource has an associated counter variable, its set and binding can be provided
+in `CounterSet` and `CounterBinding`.
+
+In the SPIR-V backend, there will be a legalization pass that will lower the
+`spirv.VulkanBuffer` type to code closer to the SPIR-V to be generated:
+
+1.  Calls to `@llvm.spv.handle.fromBinding` will be replaced by two calls. One
+    that returns a handle to the array, and another that return a handle to the
+    counter, if necessary.
+2.  Calls to `@llvm.spv.resource.getpointer` will have the handle replaced by
+    the handle of the array.
+3.  Calls to `@llvm.spv.resource.updatecounter` will be replaced by a call to
+    `@llvm.spv.resource.getpointer` with the handle of the counter followed by
+    an atomic add.
+4.  If the type of the original handle is rasterizer ordered, all uses of
+    `@llvm.spv.resource.getpointer` will be surrounded by instructions to begin
+    and end the critical region.
+
+A separate legalization pass will then move the critical region markers so that
+they follow the rules required by the SPIR-V specification. This will be the
+same as the
+[`InvocationInterlockPlacementPass`](https://github.com/KhronosGroup/SPIRV-Tools/blob/682bcd51548e670811f1d03511968bb59a1157ce/source/opt/invocation_interlock_placement_pass.h)
+pass in SPIR-V Tools.
+
+The types for the buffers must have an
+[explicit layout](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#ExplicitLayout).
+The layout information will be obtained from the DataLayout class:
+
+1.  Struct offsets will come from `DataLayout::getStructLayout`, which returns
+    the offset for each member.
+2.  The array stride will be the size of the array elements. This assumes that
+    structs have appropriate padding at the end to ensure its size is a multiple
+    of its alignment.
+3.  Matrix stride?
+4.  Row major vs Col major?
```

Comment:
I need to spend some time soon figuring out a design and plan for matrix types. There is a lot of complication there to figure out which will impact both DXIL and SPIRV.